// statigen/dist/files/EjsFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EjsFile = void 0;
const TextFile_1 = require("./TextFile");
const fsExtra = require("fs-extra");
const ejs = require("ejs");
const ejsLint = require("ejs-lint");
const util_1 = require("../util");
const DiagnosticMessages_1 = require("../DiagnosticMessages");
class EjsFile extends TextFile_1.TextFile {
    constructor(srcPath, outPath) {
        super(srcPath, outPath.replace(/\.ejs$/i, '.html'));
    }
    renderAsTemplate(file, content) {
        var _a, _b;
        const data = {
            attributes: Object.assign({}, ((_a = file.attributes) !== null && _a !== void 0 ? _a : {})),
            content: content,
            file: file,
            project: this.project,
            require: require,
            /**
             * Resolves the url (relative to template file) based on the relative position to the host file
             */
            url: (url) => {
                return (0, util_1.getRelativeUrl)(url, this.outPath, file.outPath);
            }
        };
        try {
            return ejs.render(this.text, data);
        }
        catch (e) {
            const error = e;
            //scan the file with ejs-lint to figure out the actual syntax errors
            const lintError = ejsLint(this.text, data);
            if (lintError) {
                //add diagnostic for syntax error
                file.diagnostics.push(Object.assign(Object.assign({ file: this, range: (0, util_1.createRange)(lintError.line - 1, lintError.column - 1, lintError.line - 1, lintError.column) }, DiagnosticMessages_1.DiagnosticMessages.genericError(lintError.message)), { relatedInformation: [{
                            file: file,
                            message: 'Source file'
                        }] }));
                //if this looks like an ejs error, try to extract the line and message information
            }
            else if ((_b = error.message) === null || _b === void 0 ? void 0 : _b.includes('>>')) {
                const ejsError = (0, util_1.getEjsError)(error);
                //push the underlying error...better than nothing
                file.diagnostics.push(Object.assign(Object.assign({ file: this }, DiagnosticMessages_1.DiagnosticMessages.genericError(ejsError.message)), { range: ejsError.line !== undefined ? (0, util_1.createRange)(ejsError.line - 1, 0, ejsError.line - 1, 0) : undefined, relatedInformation: [{
                            file: file,
                            message: 'Source file'
                        }] }));
                //push the underlying error...better than nothing
            }
            else {
                file.diagnostics.push(Object.assign(Object.assign({ file: this }, DiagnosticMessages_1.DiagnosticMessages.genericError(error.message)), { relatedInformation: [{
                            file: file,
                            message: 'Source file'
                        }] }));
            }
        }
    }
    publish() {
        var _a;
        fsExtra.outputFileSync(this.outPath, (_a = this.renderAsTemplate(this, '')) !== null && _a !== void 0 ? _a : '');
    }
}
exports.EjsFile = EjsFile;
