// statigen/dist/files/MarkdownFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkdownFile = void 0;
const TextFile_1 = require("./TextFile");
const marked_1 = require("marked");
const fsExtra = require("fs-extra");
const util_1 = require("../util");
class MarkdownFile extends TextFile_1.TextFile {
    constructor(srcPath, outPath) {
        super(srcPath, outPath.replace(/\.md$/i, '.html'));
    }
    /**
     * Get the title of this file (generally used for html page title and tree view)
     */
    get title() {
        var _a, _b;
        //use the title from attributes if available
        if ((_a = this.attributes) === null || _a === void 0 ? void 0 : _a.title) {
            return this.attributes.title;
        }
        return (_b = this.getH1Text()) !== null && _b !== void 0 ? _b : (0, util_1.getTitleFromFilePath)(this.outPath);
    }
    /**
     * If the markdown code starts with a h1 header, return its text
     */
    getH1Text() {
        var _a;
        const [, h1Text] = (_a = /^#\s+(.+)/.exec(this.text.trim())) !== null && _a !== void 0 ? _a : [];
        return h1Text;
    }
    publish() {
        marked_1.marked.use({
            renderer: {
                heading: ((text, level, raw, slugger) => {
                    const id = slugger.slug(text);
                    return `<a href="#${id}"><h${level} id="${id}">${text}</h${level}></a>`;
                })
            }
        });
        let html = (0, marked_1.marked)(this.text).trim();
        //remove trailing newline
        fsExtra.outputFileSync(this.outPath, this.project.generateWithTemplate(this, html));
    }
}
exports.MarkdownFile = MarkdownFile;
