// statigen/dist/files/TextFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextFile = void 0;
const fsExtra = require("fs-extra");
const frontMatter = require("front-matter");
const util_1 = require("../util");
class TextFile {
    constructor(srcPath, outPath) {
        this.srcPath = srcPath;
        this.outPath = outPath;
        /**
         * A collection of diagnostics for this file
         */
        this.diagnostics = [];
    }
    get lines() {
        if (!this._lines) {
            this._lines = this.text.split(/\r?\n/g);
        }
        return this._lines;
    }
    /**
     * Get the title of this file (generally used for html page title and tree view)
     */
    get title() {
        var _a;
        //use the title from attributes if available
        if ((_a = this.attributes) === null || _a === void 0 ? void 0 : _a.title) {
            return this.attributes.title;
        }
        return (0, util_1.getTitleFromFilePath)(this.outPath);
    }
    /**
     * Load the file from disk and parse any frontmatter
     */
    load() {
        const text = fsExtra.readFileSync(this.srcPath).toString();
        //parse any detected frontmatter
        const content = frontMatter(text);
        this.text = content.body;
        this.attributes = content.attributes;
    }
    /**
     * Get the line of source code at the specified line index, or null if the index is invalid or out of bounds
     */
    getLine(index) {
        return this.lines[index];
    }
    publish() {
        var _a;
        fsExtra.outputFileSync(this.outPath, (_a = this.project.generateWithTemplate(this, this.text)) !== null && _a !== void 0 ? _a : '');
    }
}
exports.TextFile = TextFile;
