// statigen/dist/Tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SortMode = exports.Tree = void 0;
class Tree {
    constructor(name, path, _title, file, children = []) {
        var _a;
        this.name = name;
        this.path = path;
        this._title = _title;
        this.file = file;
        this.children = children;
        this._title = (_a = this._title) !== null && _a !== void 0 ? _a : this.name;
    }
    get hasChildren() {
        var _a;
        return ((_a = this.children) === null || _a === void 0 ? void 0 : _a.length) > 0;
    }
    add(path, file) {
        var _a;
        const parts = path.split(/[\\\/]/);
        let node = this;
        const filename = parts.pop();
        for (let i = 0; i < parts.length; i++) {
            const part = parts[i];
            const partPath = parts.slice(0, i + 1).join('/');
            const search = node.children.find(x => x.name === part);
            if (search) {
                node = search;
            }
            else {
                const newNode = new Tree(part, partPath, part);
                node.children.push(newNode);
                node = newNode;
            }
        }
        //now that we have the parent node, push the file
        node.children.push(new Tree(filename, path, (_a = file === null || file === void 0 ? void 0 : file.title) !== null && _a !== void 0 ? _a : filename, file));
    }
    /**
     * Sort the tree recursively. Leafs are sorted to the top and then alphabetized, then branches are alphatized next (at the bottom)
     */
    sort() {
        var _a;
        const nodes = [this];
        while (nodes.length > 0) {
            const node = nodes.pop();
            nodes.push(...(_a = node.children) !== null && _a !== void 0 ? _a : []);
            //sort the children of this node
            node.children.sort((a, b) => {
                var _a, _b;
                let aPriority = (_a = this.getPriority(a)) !== null && _a !== void 0 ? _a : 1000;
                let bPriority = (_b = this.getPriority(b)) !== null && _b !== void 0 ? _b : 1000;
                //sort by file priority
                if (aPriority > bPriority) {
                    return 1;
                }
                else if (bPriority > aPriority) {
                    return -1;
                }
                //now sort by name
                return a.name.localeCompare(b.name);
            });
        }
        return this;
    }
    /**
     * Given a node, find its priority. If the node doesn't have a priority, look through its direct children for a `parentPriority` attribute
     */
    getPriority(node) {
        var _a, _b;
        if (!node.hasChildren) {
            return (_a = node.file) === null || _a === void 0 ? void 0 : _a.attributes.priority;
        }
        for (const child of node.children) {
            const priorityFromChild = (_b = child.file) === null || _b === void 0 ? void 0 : _b.attributes.parentPriority;
            if (priorityFromChild) {
                return priorityFromChild;
            }
        }
    }
    /**
     * Given a node, find its priority. If the node doesn't have a priority, look through its direct children for a `parentPriority` attribute
     */
    get title() {
        var _a, _b, _c;
        if (this.hasChildren) {
            for (const child of this.children) {
                const titleFromChild = (_a = child.file) === null || _a === void 0 ? void 0 : _a.attributes.parentTitle;
                if (titleFromChild) {
                    return titleFromChild;
                }
            }
        }
        return (_c = (_b = this.file) === null || _b === void 0 ? void 0 : _b.attributes.title) !== null && _c !== void 0 ? _c : this._title;
    }
}
exports.Tree = Tree;
var SortMode;
(function (SortMode) {
})(SortMode = exports.SortMode || (exports.SortMode = {}));
